File: 프로그래머스/1/136798. 기사단원의 무기/기사단원의 무기.java
class Solution {
    public int solution(int number, int limit, int power) {
        int totalScore = 0;
        
        // 각 숫자의 약수 개수를 미리 계산하여 배열에 저장
        int[] divisorCounts = new int[number + 1];
        for (int i = 1; i <= number; i++) {
            for (int j = i; j <= number; j += i) {
                divisorCounts[j]++;
            }
        }
        
        // 약수 개수를 이용하여 결과 계산
        for (int i = 1; i <= number; i++) {
            if (divisorCounts[i] > limit) {
                totalScore += power;
            } else {
                totalScore += divisorCounts[i];
            }
        }
        
        return totalScore;
    }
}

File: 프로그래머스/1/136798. 기사단원의 무기/기사단원의 무기.py
def solution(number, limit, power):
    answer = 0
    
    for i in range(1, number + 1):
        count = get_divisors_optimized(i)
        answer += count if count <= limit else power
    
    return answer

def get_divisors_optimized(n):
    divisors = set()
    for i in range(1, int(n ** 0.5) + 1):
        if n % i == 0:
            divisors.add(i)
            divisors.add(n // i)
    return len(sorted(divisors))